### DISCUSSION TITLE
use sublanguage

### DISCUSSION THREAD

**Discussion by FrankSAURET** [microsoft/vscode-discussions#738] [Extension Development QnA] [answered]
I would like to enable an extension only if a file contains a particular line. My extension concerns the microbit and I would like it to activate only if the line "from microbit import" is present. Is it possible and if so how ?

**Comment by alefragnani** [comment, marked as answer]
Hi @FrankSAURET ,

You can’t control the _activation_ of an extension this way, because it is not a supported scenario (https://code.visualstudio.com/api/references/activation-events).  I suppose the nearest activation event for your scenario would be `onLanguage`, which would limit your extension activation when you open a microbit related file (if there is one)

But, even if your extension is active, you still can enable/disable its commands. You will just have to deal with it manually, like detecting when the active file and file content changes, depending on your needs. And, another great API to help you with it will be the [when clauses](https://code.visualstudio.com/api/references/when-clause-contexts), which allows you to enable/disable commands _kind of automatically_.

Hope this helps

**Comment by FrankSAURET** [comment]
Thanks i will try this.

**Comment by FrankSAURET** [comment]
I tried but I couldn't find a clause that allows to search for text in the editor. Probably the 'in' function but I don't know how to retrieve the content from the editor that has the focus.
Can you help me ?

**Reply by r3m0t** [reply]
`vscode.window.activeTextEditor.document.lineAt()` / `vscode.window.activeTextEditor.document.lineCount`?

**Reply by FrankSAURET** [reply]
Thank's but I can't use this in a when clause ?

**Reply by alefragnani** [reply]
Hi @FrankSAURET ,

To be able to use in `when` clauses, you must use the `vscode.commands.executeCommand("setContext",` API, which is responsible to _connect_ the when clause you used in `package.json` and your code.

For instance, let's say you defined in your `package.json` something like:


```json
   "command": "myExtension.SomeCommand",
   "when": "myExtension.hasMyDesiredText"
```

Where, `hasMyDesiredText` is a Boolean attribute. 

Then, in your extension's source you could do something like

```ts
   var hasMyDesiredText = someFunctionToParseTheCurrentText();
   vscode.commands.executeCommand("setContext", myExtension.hasMyDesiredText);
```

This snippet could be fired when you switch between documents (`onDidChangeActiveTextEditor`), or whenever its content changes (`onDidChangeTextDocument`), or whenever you feel necessary.

You could use [this PR](https://github.com/alefragnani/vscode-bookmarks/pull/616) I made on my Bookmarks extension, where I add a new when clause support with custom contexts, similar to what you are trying to achieve.

Hope this helps
